### PR TITLE
deals-top-stats

### PR DESCRIPTION
**Pull request by mabbus**

### CODE REVIEW

File: app/Http/Controllers/Web/EmployeesDealsController.php [hunk 1]
```diff
@@ -54,21 +54,25 @@ public function index(EmployeeGetDeals $request)
                 return $q->where('status', $id);
             });
 
-        $amount = $deals->with('fees')->select('id', 'rate',  'amount')->get()->sum('totalAmount') ?? 1;
-        $rate = $deals->avg('rate') ?? 1;
-        $terms = $deals->avg('terms') ?? 1;
-
+        $amount = $deals->sum('amount') ?? 0;
+        $payback = $deals->get()->sum(function ($item) {
```
Comment (lexod) [inline app/Http/Controllers/Web/EmployeesDealsController.php:58]: Choose only deals columns which you actually need
Same for #63

File: app/Http/Tasks/Syndicate/GetTopSyndicateStats.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    private $syndicate;
+
+    public function __construct(Syndicate $syndicate)
+    {
+        $this->syndicate = $syndicate;
+    }
+
+    public function run()
+    {
+        $deals = $this->syndicate->load('deals')->deals;
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/GetTopSyndicateStats.php:25]: Choose only deals columns which you actually need

File: app/Http/Tasks/Syndicate/GetTopSyndicateStats.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        $payback = $deals->reduce(function($carry, $item) {
+            return $carry = $carry + ($item->pivot->amount * $item->rate);
+        },0);
+
+        $fees = $deals->reduce(function($carry, $item) {
+            return $carry = $carry + (($item->pivot->amount * $item->rate) * ($item->pivot->management_fee/100));
+        },0);
+
+        $netPayback = $payback - $fees;
+        $transactions = Transaction::where('type', Transaction::TYPE_SYNDICATE_CREDIT)
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/GetTopSyndicateStats.php:39]: Choose only deals columns which you actually need

File: app/Http/Tasks/Syndicate/GetTopSyndicateStats.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            ->when(data_get($this->syndicate->paymentProfile, 'id',0), function ($t, $id) {
+                $t->where('payment_account_id', $id);
+            });
+
+        $totalGrossCollected = $transactions->get()->sum(function ($item) {
+            return $item->amount + $item->fee;
+        });
+
+        $netCollected = $transactions->get()->sum('amount');
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/GetTopSyndicateStats.php:48]: Just ->sum() will do just fine

**Review by lexod — CHANGES_REQUESTED** [review]

File: app/Http/Tasks/Syndicate/GetTopSyndicateStats.php (hunk 4 above)
Comment (mabbus) [reply app/Http/Tasks/Syndicate/GetTopSyndicateStats.php:48]: It breaks when I remove get()

File: app/Http/Tasks/Syndicate/GetTopSyndicateStats.php (hunk 3 above)
Comment (mabbus) [reply app/Http/Tasks/Syndicate/GetTopSyndicateStats.php:39]: why? i need to sum up transactions?

Comment (lexod) [reply app/Http/Tasks/Syndicate/GetTopSyndicateStats.php:39]: Why do you need created_at/updated_at for example? All you need are id, type, amount and deal_id

File: app/Http/Tasks/Syndicate/GetTopSyndicateStats.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    public function __construct(Syndicate $syndicate)
+    {
+        $this->syndicate = $syndicate;
+    }
+
+    public function run()
+    {
+        $deals = $this->syndicate->load('deals')->deals;
+        $amount = $deals->reduce(function($carry,$item) {
+            return $carry = $carry + $item->pivot->amount;
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/GetTopSyndicateStats.php:27]: You don't need carry assignment in any of these rows. Just return carry + YOUR EXPRESSION

**Review by lexod — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Web/EmployeesDealsController.php [hunk 6]
```diff
@@ -55,21 +55,25 @@ public function index(EmployeeGetDeals $request)
...
 
-        $amount = $deals->with('fees')->select('id', 'rate',  'amount')->get()->sum('totalAmount') ?? 1;
-        $rate = $deals->avg('rate') ?? 1;
-        $terms = $deals->avg('terms') ?? 1;
-
+        $amount = $deals->sum('amount') ?? 0;
+        $payback = $deals->get(['amount','rate'])->sum(function ($item) {
+            return $item->payback;
+        });
+        $rate = $payback / $amount;
```
Comment (lexod) [inline app/Http/Controllers/Web/EmployeesDealsController.php:62]: 0 division exception if $amount === 0

File: app/Http/Controllers/Web/EmployeesDealsController.php [hunk 7]
```diff
@@ -55,21 +55,25 @@ public function index(EmployeeGetDeals $request)
...
             'rate' => $rate,
-            'buy_rate' => $deals->avg('buy_rate'),
-            'payback' => ($rate * $amount),
-            'terms' => $terms,
-            'term_type' => $deals->avg('term_type'),
-            'closing_fee' => $deals->avg('closing_fee'),
-            'payment' => $amount / $terms,
-            'commissionAttribute' => 1,
-            'progress' => '',
+            'payment' => $deals->get()->sum(function($item) {
```
Comment (lexod) [inline app/Http/Controllers/Web/EmployeesDealsController.php:71]: duplicate query. create a collection you will be applying sums to
same for 64

**Review by lexod — CHANGES_REQUESTED** [review]

**Review by lexod — APPROVED** [review]

**Pull request closed without merging**